### PR TITLE
iso_c_binding: add debug support.

### PR DESCRIPTION
**Pull request by fghoussen**
This should fix issue #82

### CODE REVIEW

**Comment by fghoussen** [comment]
CI was OK at my side ?! No time for this today, I'll try to fix / check it this week-end. Would say if you dont call debug_c, there is no debug (init already OK). Not sure. Will check

**Comment by davydden** [comment]
> Would say if you dont call debug_c, there is no debug (init already OK)

Form my experience, no. Note that you can't easily check this as https://github.com/opencollab/arpack-ng/issues/82 happened on HPC cluster whereas I also don't have any problems / debug output on my mac or Ubuntu. 
As I mentioned in https://github.com/opencollab/arpack-ng/issues/82 , simply patching source to disable all output whatsoever fixed it on a cluster.

**Comment by fghoussen** [comment]
CI is OK at my side !?... Weird. I'll check it when possible

**Comment by fghoussen** [comment]
Would say CI fails as coverage go from 70% to 69% ! :D I'll check tomorrow

**Comment by fghoussen** [comment]
I've pushed comestic changes that should illustrate things the way you (@davydden) probably want them to be : hope this will get the coverage back from 69.5 to 70% !... :D 

My understanding is that as long as you do not initialize the common debug block (at Fortran side), this block will be considered to be in an "uninitialized state" which stands for "no debug state" (actually, I believe this happens already in 99.99% of cases where you don't need/care about debug traces). Look at the icb examples : arpack is ran twice, the 1st time is performed the classical way (no debug), then you set debug flags (`debug_c`), then you run arpack a 2d time and get debug output (according to flags).

@davydden : does this sounds good to you ? I could also add an icb method like `init_arpack_c`, but, as I can't call that implicitly for arpack users (where to do this ?), any user will have to call it (always) explicitly : no benefit when compared to the "calling `debug_c` approach". Also, the common debug block could be initialized to zero in Fortran, but, I have to say I have no idea how / where to do that !...

Franck

```
~> icb_arpack_c
992.000000
993.000000
994.000000
995.000000
996.000000
997.000000
998.000000
999.000000
1000.000000
------

 _getv0: B-norm of initial / restarted starting vector
 -----------------------------------------------------
    1 -    1:    2.09029D+04
  

 _getv0: B-norm of initial / restarted starting vector
 -----------------------------------------------------
    1 -    1:    2.07939D+04
  
1000.000000 1000.000000
999.000000 999.000000
998.000000 998.000000
997.000000 997.000000
996.000000 996.000000
995.000000 995.000000
994.000000 994.000000
993.000000 993.000000
992.000000 992.000000



```

**Comment by fghoussen** [comment]
By the way

>Consequently, I am quite certain the issue **is** there, but for whatever reason it does not appear on all platforms/compilers/os.

I believe this is a direct consequence of lack of Fortran specifications (<=> "uninitialized state" of the common block) : this probably could be OS/compiler/implementation dependent... No great, but, this is Fortran world. I would say the "preferred" solution  here would be to use `-finit-real=zero -finit-integer=0` at compile time : @davydden , did you try ?

**Comment by fghoussen** [comment]
youpi ....

**Comment by fghoussen** [comment]
@davydden : Seems the last commit (64f3c1b) should fix your problem on your HPC cluster whatever compiler options... Not sure. Does not break anything apparently at my side.... And this is supposed to initialize the debug common block at fortran side : check it and you know ! (I'am not sure - can't test it)

**Comment by davydden** [comment]
> My understanding is that as long as you do not initialize the common debug block (at Fortran side), this block will be considered to be in an "uninitialized state" which stands for "no debug state"

that's not what folks here https://stackoverflow.com/a/49618639/888478 say. I think it's exactly the problem -- this is somehow os/compiler/platform dependent. I have no output on macOS or Ubuntu, but things have broken on HPC cluster.

> -finit-real=zero -finit-integer=0 at compile time

I have not because folks at stackoverflow https://stackoverflow.com/a/49618639/888478 said those flags do not influence COMMON block variables ;-(

> Seems the last commit (64f3c1b) should fix your problem on your HPC cluster

I certainly will! Thanks for not giving up on this issue. 👍

**Comment by fghoussen** [comment]
Seems initialization of debug common block can not be moved to `debug.h` (compil KO) : I reverted commit 22d1b3b !... This means initialization of this common block will be done **only** using icb at configure time...

@sylvestre : if you like this, you may fixup all this misery in one relevant commit !... :D

File: debug_icb.f90 [hunk 1]
```diff
@@ -37,3 +37,16 @@ subroutine debug_c(logfil_c, ndigit_c, mgetv0_c,
   mcgets = mcgets_c
   mceupd = mceupd_c
 end subroutine debug_c
+
+! Initialisation of the debug common block to "no debug".
+block data init
```
Comment (davydden) [inline debug_icb.f90]: can you move this into a separate new `.f90`? It should not be related to ICB and should happen regardless of users having ICB on or off. 

Currently, this source file is not part of the lib unless ICB is on.

File: debug_icb.f90 [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  mcaup2 = mcaup2_c
+  mcaitr = mcaitr_c
+  mceigh = mceigh_c
+  mcapps = mcapps_c
+  mcgets = mcgets_c
+  mceupd = mceupd_c
+end subroutine debug_c
+
+! Initialisation of the debug common block to "no debug".
+block data init
```
Comment (davydden) [inline debug_icb.f90]: i would even move the next 10 lines into a stand-alone PR in a separate file like `init_common.f90`. Then if it works (i will check), the changes are un-controversial and i think @sylvestre can merge it quickly.

Comment (sylvestre) [reply debug_icb.f90]: sure, just ping me when it is ready!

**Comment by fghoussen** [comment]
Pushing from McDonald highway !... If KO, will get back to it next week. If OK, for me, good to go.

**Comment by sylvestre** [comment]
Travis is happy, improvement of the code coverage, taking it.
We can always land follow up commits later.
bravo & thanks

**Comment by fghoussen** [comment]
@davydden : test it on your HPC cluster if you have some time !... This is a blind fix :D

**Comment by davydden** [comment]
> test it on your HPC cluster if you have some time !... This is a blind fix :D

i gave it a try, unfortunately i still see the issues that Arpack tries to print some debug info ;-(

**Comment by fghoussen** [comment]
>i gave it a try, unfortunately i still see the issues that Arpack tries to print some debug info ;-(

Did you `make distclean` or `rm -fr arpack-ng` ?... Before re-bootstrapping + configuring... If yes, then you are welcome to PR if you find the bug !

**Comment by davydden** [comment]
> Did you make distclean or rm -fr arpack-ng ?... Before re-bootstrapping + configuring...

sure, I always build in a clean environment from scratch.

**Comment by fghoussen** [comment]
Too bad ! Blind fixing is definitively not the best way to go... :D

**Comment by fghoussen** [comment]
@davydden : seems you need to "put" the data block in each subroutine or program !?... The more I deal with Fortran, the more I feel bad !

See here http://www.personal.psu.edu/jhm/f90/questions/common.html (how to use block data) and here http://www.personal.psu.edu/jhm/f90/examples/ode/fall.f

So if you feel courageous enough you can "put" that in all subroutines (*[ae]upd*), or, you find THE one initialization method that parpack/arpack may use in all cases (if possible ?). For now my understanding is that the "data block" is here but not called/used !

Hope this helps !...

**Comment by caliarim** [comment]
No, it is not necessary to put the data block in each subroutine. If you put the attached test.f in PARPACK/MPI/SRC and compile with gfortran test.f ./.libs/libarpack.a and run ./a.out you see that the data initialization works both in the program and in the subroutine. I think there is no problem at fortran level.
[test.f.gz](https://github.com/opencollab/arpack-ng/files/1893103/test.f.gz)

**Comment by davydden** [comment]
@caliarim 

> I think there is no problem at fortran level.

I think the problem is there, but I don't know why. What I know and checked, though, is that if I comment out the usage of those common-block debug variables by simply changing
```
msglvl = mgetv0
```
to
```
msglvl = 0
```
then when calling pArpack function from C++ there is no debug output (as epxected). If I don't, even with this PR merged, pArpack tries to print output and fails with 

```
At line 68 of file <blabla>/arpack-ng-3.4.0/PARPACK/UTIL/MPI/pdvout.f
Fortran runtime error: Bad unit number in statement
```

Why exactly it fails is irrelevant, the issue is that it calls printing routines which means those variables are not set to zero despite https://github.com/opencollab/arpack-ng/pull/107/files#diff-ffb8d149270531d336dde96ac0c0704eR11.

**Comment by caliarim** [comment]
The external statement is recommended in order not to forget to add the block data statement. But it is not mandatory. Did you try my test.f?

**Comment by fghoussen** [comment]
Connected with not convenient tablet !... Got an idea. Will test it and let you know tomorrow

**Comment by fghoussen** [comment]
@davydden  : try to add an empty dummy fortran program in debug_init.f90. Does it fix the hpc cluster problem ? Arpack is a lib (probably no main, just subroutines). Data block may be "loaded" before main (= fortran program) to init global variables... No main, no init ?!...

**Comment by fghoussen** [comment]
If KO, try to add " external debug_init " in the dummy prg

**Comment by fghoussen** [comment]
By the way, at my side, I tried: 
```
 >> git diff
diff --git a/debug_init.f90 b/debug_init.f90
index f4c4f8e..74ae137 100644
--- a/debug_init.f90
+++ b/debug_init.f90
@@ -10,3 +10,7 @@ block data debug_init
        mcaupd, mcaup2, mcaitr, mceigh, mcapps, mcgets, mceupd &
        / 6, -3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 /
 end block data debug_init
+
+program dummy
+  external debug_init
+end program dummy

```
This does not seem to break anything at my side but I don't know if it can fix the cluster problem !

**Comment by fghoussen** [comment]
```
+program dummy
+  include 'debug.h'
+  external debug_init
+end program dummy

```

Seems to be OK too !...

@caliarim : I've just looked at your test.f, I would say (not sure) it works _because_ there is a `program` !

**Pull request merged**